src/pytest_alembic/tests/experimental/all_models_register_on_metadata.py

```diff
@@ -0,0 +1,191 @@
+
+    Importantly, this cannot simply import the MetaData directly, as that may have already
+    had side-effects performed on it such that it's no longer representative of if one were to
+    directly import the MetaData alone.
+
+    Thus, we start a python subprocess which does the most minimal import we can perform to extract the MetaData,
+    and return the set of tables attached to it.
+    """
+
+    command = [
+        "python",
+        "-m",
+        "pytest_alembic.tests.experimental.collect_clean_alembic_environment",
+        shlex.quote(url),
+    ]
+    import os
+    try:
+        result = subprocess.run(  # nosec
+            command,
+            check=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+            env=os.environ,
+        )
+    except subprocess.CalledProcessError as e:
+        raise AlembicTestFailure(e.stderr)  # pragma: no cover
+
+    lines = result.stdout.splitlines()
+    modules = set()
+    tablenames = set()
+
+    empty_line = 0
+    for line in lines:
+        if not line:
+            empty_line += 1
+            continue
+
+        if empty_line == 0:
+            modules.add(line)
+        elif empty_line == 1:
+            tablenames.add(line)
+
+    return sorted(modules), set(tablenames)
+
+
+def get_full_tableset(*module_names: str) -> Set[str]:
+    """Get the set of full set of tables which are defined.
+
+    The theory is that if we import every module in the tree below ma.models, we should
+    cause the import side-effect of all models being attached to the model base.
+
+    Once done, we should end up with the theoretical maximum of tables included in the MetaData.
+    """
+    metadata = None
+    for module_name in module_names:
+        if metadata is not None:
+            break
+
+        module, *_ = traverse_modules(module_name)
+        for item in module.__dict__.values():
+            if isinstance(item, MetaData):
+                metadata = item
+                break
+            if isinstance(item, DeclarativeMeta):
+                metadata = item.metadata
+                break
+
+    if metadata is None:
+        raise AlembicTestFailure(
+            "Unable to locate a MetaData reference in the local context of your `env.py`. This "
+            "test requires a local reference to either a `DeclarativeMeta` (i.e. declarative model "
+            "base), or a `MetaData` to exist in your `env.py`."
+        )
+
+    tables = metadata.tables
+    tablenames = list(tables.keys())
+    return set(tablenames)
+
+
+def traverse_modules(package_name: str):
+    """Dynamically traverse the tree of packages below a given root and import them.
+
+    Note this will perform the import of each module, which is an operation which can
+    cause side-effects!
+    """
+
+    try:
+        package = importlib.import_module(package_name)
+    except ImportError:
+        return
```

Comment:
These untested lines will become easily testable if you structure for DI, e.g. by adding a keyword-only `import_module=importlib.import_module` argument and passing a mock in your tests.

Reply:
they're not actually untested, I think coverage is just not sophisticated enough to follow the e2e tests. the majority of the testing of this plugin as a whole is through pytest's testdir fixture due to how the internals of alembic work which make it very challenging to actually meaningfully test alembic without on-disk histories for alembic to run against.

Although that's a fair suggestion so that i **can** get legit coverage without e2e tests, thanks!